services/containerregistry/refreshtoken.go

```diff
@@ -0,0 +1,112 @@
+package containerregistry
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// ExchangeRefreshToken exchanges an Azure AD token for an Azure Container Registry refresh token
+func (c *ContainerRegistryClient) ExchangeRefreshToken(ctx context.Context) (string, *RefreshTokenClaims, error) {
+	token, err := c.authorizer.Token()
+	if err != nil {
+		return "", nil, err
+	}
+
+	serviceURL, err := parseService(c.serverURL)
+	if err != nil {
+		return "", nil, err
+	}
+
+	data := url.Values{}
+	data.Set("grant_type", "access_token")
+	data.Set("service", serviceURL.Hostname())
+	data.Set("access_token", token.AccessToken)
+	if len(c.tenantID) > 0 {
+		data.Set("tenant", c.tenantID)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, fmt.Sprintf("%s://%s/oauth2/exchange", serviceURL.Scheme, serviceURL.Host), strings.NewReader(data.Encode()))
+	if err != nil {
+		return "", nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := c.httpClient.Do(req)
+	if err != nil {
+		return "", nil, err
+	}
+
+	resBytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return "", nil, err
+	}
+
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return "", nil, fmt.Errorf("received non-200 status code - %d: %s", res.StatusCode, string(resBytes))
+	}
+
+	var resData struct {
+		RefreshToken string `json:"refresh_token"`
+	}
+
+	err = json.Unmarshal(resBytes, &resData)
+	if err != nil {
+		return "", nil, err
+	}
+
+	rtClaims, err := decodeRefreshTokenWithoutValidation(resData.RefreshToken)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return resData.RefreshToken, rtClaims, nil
+}
+
+// RefreshTokenClaimsPermissions contains the actions permitted for the refresh token
+type RefreshTokenClaimsPermissions struct {
+	Actions    []string `json:"Actions"`
+	NotActions []string `json:"NotActions"`
+}
+
+// RefreshTokenClaims contians the claims of the refresh token
+// Please observe: These claims are in no way validated, only decoded
+type RefreshTokenClaims struct {
```

Comment:
Would it be better to use the [`auth.Claims`](https://github.com/manicminer/hamilton/blob/main/auth/claims.go#L12-L27) struct instead of a separate struct here? We can add more fields to it if necessary.

Reply:
I completely agree. Will look into what claims it contains and see if it's possible to reuse.

In this case, the only real value is the exp (as of now).

Reply:
Sounds good - if the claims are sufficiently semantically different between access and refresh tokens, then I agree it makes to split them. But I think we should implement this in the `auth` package. Perhaps if there is sufficient overlap we can embed the common claims.

Reply:
@manicminer I think the RefreshTokenClaims here are to specific to Container Registry to be used else where, as the `permissions` won't be implemented (most likely) in any other API. I have no problem moving it to `auth`, but I do think it may confuse others that it's in some way related to an Azure AD refresh token wich it isn't.